modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1978,13 +1980,109 @@ def sort_by(
             Column label(s) to use to determine lexicographical ordering.
         ascending : boolean, default: True
             Whether to sort in ascending or descending order.
+        **kwargs : dict
+            Keyword arguments to pass when sorting partitions.
 
         Returns
         -------
         PandasDataframe
             A new PandasDataframe sorted into lexicographical order by the specified column(s).
         """
-        pass
+        if not isinstance(columns, list):
+            columns = [columns]
+        # When we do a sort on the result of Series.value_counts, we don't rename the index until
+        # after everything is done, which causes an error when sorting the partitions, since the
+        # index and the column share the same name, when in actuality, the index's name should be
+        # None. This fixes the indexes name beforehand in that case, so that the sort works.
+
+        def sort_function(df):
+            index_renaming = None
+            if any(name in df.columns for name in df.index.names):
+                index_renaming = df.index.names
+                df.index = df.index.set_names([None] * len(df.index.names))
+            df = df.sort_values(by=columns, ascending=ascending, **kwargs)
+            if index_renaming is not None:
+                df.index = df.index.set_names(index_renaming)
+            return df
+
+        axis = Axis(axis)
+        if axis == Axis.ROW_WISE:
+            # If this df is empty, we don't want to try and shuffle or sort.
+            if len(self.axes[0]) == 0 or len(self.axes[1]) == 0:
+                return self.copy()
+            # If this df only has one row partition, we don't want to do a shuffle and sort - we can
+            # just do a full-axis sort.
+            if len(self._partitions) == 1:
+                return self.apply_full_axis(
+                    1,
+                    sort_function,
+                )
+            if self.dtypes[columns[0]] == object:
+                # This means we are not sorting numbers, so we need our quantiles to not try
+                # arithmetic on the values.
+                method = "inverted_cdf"
+            else:
+                method = "linear"
+            from .utils import build_sort_functions
+
+            shuffling_functions = build_sort_functions(
+                self,
+                columns[0],
+                method,
+                ascending if not isinstance(ascending, list) else ascending[0],
+                **kwargs,
+            )
+            major_col_partition_index = self.columns.get_loc(columns[0])
+            cols_seen = 0
+            index = -1
+            for i, length in enumerate(self._column_widths_cache):
+                cols_seen += length
+                if major_col_partition_index <= cols_seen:
+                    index = i
+                    break
```

Comment:
a) we should assert that we do find a column (easiest is to add an `else` clause to `for` loop)
b) if that is determining the partition index for said column, I thought we had some specific mapping to do so, cc @dchigarev

Reply:
We can use `_get_dict_of_block_index`, although this also computes the internal index into the partition for the specified index, which is unnecessary computation imo?

Reply:
oh, I forgot about the extra computation, maybe drop that (or measure how much we're losing?..)